Fixes part of #10415: Refactor recommendation validators

sunny1401 opened the pull request:
## Overview 

1. This PR fixes or fixes part of #10415.
2. This PR does the following: Refactor recommendations validators and put it into separate files.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.

sunny1401 committed:
refactor recommendation validators and recommendation validator test

sunny1401 committed:
fixing linting issue in refactor_validator_test.py

sunny1401 committed:
Merge branch 'develop' into rcomm-ref

EricZLou commented on core/domain/recommendations_validators_test.py:
```diff
@@ -0,0 +1,332 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Unit tests for core.domain.opportunity_validators."""
```
```suggestion
"""Unit tests for core.domain.recomendations_validators."""
```

sunny1401 replied:
ohh sorry.. I will just change it.

EricZLou reviewed (changes requested):
Congrats on your first PR to Oppia! Super small suggestion is below, but otherwise looks good to me!

sunny1401 committed:
pr comment - changing from Unit tests for core.domain.opportunity_validators. to Unit tests for core.domain.recommendations_validators.

EricZLou commented:
Also just noticed that there are some tests that are failing in Backend tests! Click on "Details" next to it, and let's take a look at the output log for that. Seems like this is a part of the errors:
```
ERROR: test_model_with_last_updated_greater_than_current_time (core.domain.recommendations_validators_test.TopicSimilaritiesModelValidatorTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/oppia/oppia/core/domain/recommendations_validators_test.py", line 197, in test_model_with_last_updated_greater_than_current_time
    with self.mock_datetime_for_audit(mocked_datetime):
AttributeError: 'TopicSimilaritiesModelValidatorTests' object has no attribute 'mock_datetime_for_audit'
```

Why don't you take a hack at fixing that? Message me through this Github PR or DM me on Slack if you can use some guidance! I'm here to help : )

sunny1401 committed:
Merge branch 'rcomm-ref' of https://github.com/sunny1401/oppia into rcomm-ref

sunny1401 commented:
Yes.. I just noticed it.. I will try it out and push..I will reach out if I can't fix it. Thank you

sunny1401 committed:
fixing test cases for backend

sunny1401 commented:
@EricZLou - Hi - so I fixed the test and everything passes except Travis Ci Build. Can you help me in fixing that. I am not completely sure as to why its failing.

EricZLou commented:
Awesome! I'll take a look soon @sunny1401

sunny1401 commented:
thank you

sunny1401 committed:
Merge branch 'develop' into rcomm-ref

sunny1401 committed:
trying to fix lint error

sunny1401 committed:
trying to fix lint error - removing trailing spaces

sunny1401 committed:
trying to fix lint error - replacing ndb with datastore_services

sunny1401 committed:
trying to fix lint error - fixing line too long error

sunny1401 committed:
trying to fix lint error - lint error

sunny1401 committed:
accidentally removed the SkillModelValidator code

sunny1401 committed:
accidentally removed the SkillModelValidator code- Files must have a single newline above args in doc string.

sunny1401 commented:
@EricZLou - Hi - I think I have fixed everything.. I am not sure why is it failing.. I have fixed the merge conflicts. Could you have a look. I cannot rerun the tests.

EricZLou commented on core/domain/recommendations_validators.py:
```diff
@@ -0,0 +1,113 @@
+        models.Registry.import_models([
+            models.NAMES.exploration,
+            models.NAMES.skill, models.NAMES.story,
+            models.NAMES.topic, models.NAMES.recommendations]))
```
Don't think we need `skill_models`, `story_models`, or `topic_models`! You can remove them from this code block.
```suggestion
exp_models, recommendations_models = (
   models.Registry.import_models([
      models.NAMES.exploration, models.NAMES.recommendations]))
```

EricZLou commented on core/domain/recommendations_validators_test.py:
```diff
@@ -0,0 +1,333 @@
+
+datastore_services = models.Registry.import_datastore_services()
+
+(
+    exp_models, skill_models, story_models,
+    topic_models, recommendations_models) = (
+        models.Registry.import_models([
+            models.NAMES.exploration,
+            models.NAMES.skill, models.NAMES.story,
+            models.NAMES.topic, models.NAMES.recommendations]))
```
Don't think we need `skill_models`, `story_models`, or `topic_models`! You can remove them from this code block.
```suggestion
exp_models, recommendations_models = (
   models.Registry.import_models([
      models.NAMES.exploration, models.NAMES.recommendations]))
```

EricZLou reviewed (approved):
Sometimes the tests fail randomly -- I just restarted the tests. Let's see if they pass now!

sunny1401 committed:
removing additional moels

sunny1401 committed:
fixing Imports are incorrectly sorted

sunny1401 committed:
fixing Imports are incorrectly sorted

sunny1401 commented:
Hi - so I tried changing the files.. but not there is an error in googleappengine. I feel maybe that's why the build here is failing. Any idea how to fix that bit?

DubeySandeep reviewed (approved):
LGTM for the codeowner files.

EricZLou commented:
@sunny1401 I'll restart the tests a few more times. You should be fine here!

EricZLou commented:
@aks681 PTAL as a code owner -- thanks!

DubeySandeep commented:
@sunny1401, Can you please resolve the merge conflict on this PR?  

@aks681, CAn you please review this PR?

sunny1401 committed:
Merge branch 'develop' into rcomm-ref

sunny1401 committed:
Merge branch 'rcomm-ref' of https://github.com/sunny1401/oppia into rcomm-ref

sunny1401 committed:
resolving conflicts correctly

sunny1401 commented:
@DubeySandeep - I have resolved the conflicts. I would be grateful if this could be reviewed.

sunny1401 commented:
@EricZLou @DubeySandeep Could one of review check the error/restart the tests please. I resolved the conflicts but because so much has changed in develop.. I am not sure if the errors are because of any mistake I made.

EricZLou commented:
You should be good! I'll re-run the failing tests. Sometimes, they fail randomly out of your control

sunny1401 committed:
Merge branch 'develop' into rcomm-ref

sunny1401 commented:
@aks681 Hi - could you please review this. I have removed latest the conflicts.

aks681 commented on core/domain/recommendations_validators.py:
```diff
@@ -0,0 +1,111 @@
+        exploration ids.
+
+        Args:
+            item: datastore_services.Model. ExplorationRecommendationsModel
```
Is this an `ExplorationRecommendationsModel` instance, if so have that as type here.

sunny1401 replied:
hi - I am not sure what you mean .. do you mean adding typing to the function. Cause I thought item: datastore_services.Model. ExplorationRecommendationsModel in args gives more info. Could you elaborate please @aks681

EricZLou replied:
Seems like all of the validator classmethods in `core/domain/prod_validators.py` follow this format for the Args:
```
item: datastore_services.Model. xModel to validate.
```
Maybe we should keep this block of code the way it is unless we want to do a uniform change across all of the validator classmethods in a later PR? @aks681

aks681 replied:
Oh, the whole thing represents the model? It's fine then. Because of the space, I thought the second part was the description.

aks681 commented on core/domain/recommendations_validators.py:
```diff
@@ -0,0 +1,111 @@
+        topics = list(item.content.keys())
+        data = '%s\n' % ','.join(topics)
+
+        for topic1 in topics:
+            similarity_list = []
+            for topic2 in item.content[topic1]:
```
Rename topic1 and topic2 to something more meaningful? Like `topic` and `topic_to_compare` maybe? not sure of the context here.

sunny1401 replied:
done. @aks681

aks681 reviewed (commented):
Actually, 2 comments, else lgtm!

sunny1401 committed:
pr -comments.. renaming variables in a loop

sunny1401 committed:
linting error

sunny1401 commented:
@EricZLou - could you please restart the tests. @ fail - I am not sure why? Also @aks681 - could you please elaborate your comments.. so I can fix it and then this pr can be closed or moved forward?

aks681 reviewed (approved):
Lgtm as codeowner

sunny1401 committed:
Merge branch 'develop' into rcomm-ref

sunny1401 committed:
Update prod_validators_test.py

Removing unused import json - lint error

sunny1401 committed:
Update prod_validators.py

Removing unused import - linting error

sunny1401 committed:
Merge branch 'develop' into rcomm-ref

EricZLou commented:
Hey @sunny1401 thanks for continuing to work on this! #10415 has a lot of active branches at the same time, and that's why we get the "HAS MERGE CONFLICTS" notice so often -- other contributors are working on the same files as you. Once you fix the small lint check errors, I'll re-run the tests until they all pass, and we'll get this PR merged!

sunny1401 committed:
Update prod_validators.py

fixing lint error in prod_validators.py

sunny1401 committed:
Update prod_validators_test.py

fixing lint_error in prod_validators_test.py

sunny1401 commented:
@EricZLou - sorry for the delay. Have fixed the error

sunny1401 commented:
@EricZLou 
Could you please restart the tests?

EricZLou commented:
@sunny1401 thanks for the ping -- done!
